frame/merkle-mountain-range/src/mmr/storage.rs

```diff
@@ -133,23 +132,22 @@ where
 		// Effectively move a rolling window of fork-unique leaves. Once out of the window, leaves
 		// are "canonicalized" in offchain by moving them under `Pallet::node_canon_offchain_key`.
 		let leaves = NumberOfLeaves::<T, I>::get();
-		let window_size =
-			<T as frame_system::Config>::BlockHashCount::get().unique_saturated_into();
+		let window_size = Pallet::<T, I>::offchain_canonicalization_window();
```

Comment:
Am I right that we never prune non-canonical (`node_offchain_key`) offchain db entries that are inserted during initial sync? Then shall we care about x2 offchain db growth on fresh nodes? Maybe we shall add more pruning here, or it doesn't worth it, wdyt?

Reply:
That's right, during initial sync there is no offchain db pruning because offchain worker isn't run.

Afaict there is no way for runtime to know whether imported block was during initial sync or whether it's already been finalized, etc; and we have no other mmr component running outside of runtime that could prune these.

----

I will follow-up with a PR that tries to do this extra pruning as follows:
- change fork-resistant key to `(prefix, pos, parent_hash)` and hope (I need to verify) there is functionality to get all entries matching `(prefix, pos)` partial key,
- keep offchain db entry `last_pruned_leaf_number`
- in offchain worker attempt to prune `(last_pruned_leaf_number, current - window]` rather than just `current - window`.
- we can even do this in batches; put a limit on how many to clean up in each offchain_worker invocation so that we spread the load over time.

WDYT?

Reply:
Yeah, that looks good to me

Reply:
Opened https://github.com/paritytech/substrate/issues/12508 to track this